### PR TITLE
KEYCLOAK-17501 Add support for map storage in WildFly

### PR DESCRIPTION
**Pull request by hmlnarik**
Support for _map_ storage in WildFly server subsystem. Draft as the contained CLI script will be updated after the following PRs are merged:
- [x] #7842
- [x] #7844
- [x] #7878

### CODE REVIEW

File: distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+/subsystem=keycloak-server/spi=authenticationSessions:add(default-provider=map)
+/subsystem=keycloak-server/spi=client:add(default-provider=map)
+/subsystem=keycloak-server/spi=clientScope:add(default-provider=map)
+/subsystem=keycloak-server/spi=group:add(default-provider=map)
+# /subsystem=keycloak-server/spi=realm:add(default-provider=map)
+/subsystem=keycloak-server/spi=role:add(default-provider=map)
+/subsystem=keycloak-server/spi=user:add(default-provider=map)
+
+/subsystem=keycloak-server/spi=mapStorage:add(default-provider=concurrenthashmap)
+/subsystem=keycloak-server/spi=mapStorage/provider=concurrenthashmap:add(properties={dir="${jboss.server.data.dir}/map"},enabled=true)
```
Comment (mhajas) [inline distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli:12]: We should disable the UserFederation provider.

File: distribution/server-dist/assembly.xml [hunk 2]
```diff
@@ -118,6 +118,13 @@
                 <include>licenses-${product.slot}/**</include>
             </includes>
         </fileSet>
+        <fileSet>
+            <directory>src/main/docs</directory>
+            <outputDirectory>docs</outputDirectory>
+            <includes>
+                <include>**</include>
+            </includes>
+        </fileSet>
```
Comment (vramik) [inline distribution/server-dist/assembly.xml:127]: Can you please explain why we need this change? Thank you

File: distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli [hunk 3]
```diff
@@ -0,0 +1,14 @@
+embed-server
+
+/subsystem=keycloak-server/spi=authenticationSessions:add(default-provider=map)
+/subsystem=keycloak-server/spi=client:add(default-provider=map)
+/subsystem=keycloak-server/spi=clientScope:add(default-provider=map)
+/subsystem=keycloak-server/spi=group:add(default-provider=map)
+# /subsystem=keycloak-server/spi=realm:add(default-provider=map)
```
Comment (vramik) [inline distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli:7]: Now we can un-comment it, cannot we?

**Review by vramik — COMMENTED** [review]
Thank you @hmlnarik for the PR, it looks good to me, I'm aware it's still draft PR, I'm providing the "draft" review for now and I plan to get back to it later. :)

File: distribution/server-dist/assembly.xml (hunk 2 above)
Comment (hmlnarik) [reply distribution/server-dist/assembly.xml:127]: yes, that's correct

File: distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-model-map/main/module.xml [hunk 4]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2016 Red Hat, Inc. and/or its affiliates
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-model-map/main/module.xml]: yup

File: distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli (hunk 3 above)
Comment (hmlnarik) [reply distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli]: done

File: distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli (hunk 1 above)
Comment (hmlnarik) [reply distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli:33]: Looking through the code, it's not that easy. Disabling `jpa` provider leads to an exception on startup, so I'd rather keep it as is - map storage it is a preview feature and the user federation will be updated to tree store soon anyway.

**Review by vramik — COMMENTED** [review]
The PR LGTM, thank you @hmlnarik. There are two failing tests with `auth-server-wildfly` (with cli sript applied): 

- `DeployedScriptAuthenticatorTest` - there is an issue with non-UUID string during import
-  `WelcomePageTest` - which should be skipped because it eraseses database and the database is not used 

EDIT: Both issues are test related and if we doesn't require testsuite to pass it can go in IMO.

**Comment by hmlnarik** [comment]
> The PR LGTM, thank you @hmlnarik. There are two failing tests with `auth-server-wildfly` (with cli sript applied):
> 
> * `DeployedScriptAuthenticatorTest` - there is an issue with non-UUID string during import
> * `WelcomePageTest` - which should be skipped because it eraseses database and the database is not used
> 
> EDIT: Both issues are test related and if we doesn't require testsuite to pass it can go in IMO.

Thank you, good catch! I believe they should pass before this can be merged. The failures will be fixed by setting the proper key type similarly to [here](https://github.com/keycloak/keycloak/blob/b43056abb2e325d316de07bb32fb0095765bc4e2/testsuite/integration-arquillian/tests/base/src/test/resources/META-INF/keycloak-server.json#L86-L87) once #7952 would be in.

**Comment by hmlnarik** [comment]
> > The PR LGTM, thank you @hmlnarik. There are two failing tests with `auth-server-wildfly` (with cli sript applied):
> > 
> > * `DeployedScriptAuthenticatorTest` - there is an issue with non-UUID string during import
> > * `WelcomePageTest` - which should be skipped because it eraseses database and the database is not used
> > 
> > EDIT: Both issues are test related and if we doesn't require testsuite to pass it can go in IMO.
> 
> Thank you, good catch! I believe they should pass before this can be merged. The failures will be fixed by setting the proper key type similarly to [here](https://github.com/keycloak/keycloak/blob/b43056abb2e325d316de07bb32fb0095765bc4e2/testsuite/integration-arquillian/tests/base/src/test/resources/META-INF/keycloak-server.json#L86-L87) once #7952 would be in.

Actually, thinking about it a bit more, I believe we can reverse the two PRs and merge this first. We know that this will be fixed with #7952, the problem is in tests, the fix is on the way, and there is no automation running (and blocking) the PR. WDYT, @mhajas , @vramik?

File: distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+## Apply this file using the following command from the Keycloak root directory:
+##
+##   bin/jboss-cli.sh --file=docs/examples/map-storage-concurrenthashmap.cli
+##
+## This will modify standalone/configuration/standalone.xml
+##
+
+embed-server
+
+/subsystem=keycloak-server/spi=authorizationPersister:add(default-provider=map)
```
Comment (mhajas) [inline distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli:16]: Should we add a possibility to configure default-provider using env variable similarly as we have for undertow?

Something like:
```suggestion
/subsystem=keycloak-server/spi=authorizationPersister:add(default-provider=${keycloak.authorization.provider:map})
```

This way it is easier to use map storage only for `userSessions` for example.

**Review by mhajas — APPROVED** [review]
Thanks @hmlnarik, this looks great. I tested locally and it works. I would say it is ok to merge even though these two tests are failing as it is not problem in map storage and we are sure this will be fixed in the next version. 

Should we add something also to the documentation/release notes? 

I added one comment into the code but I don't think it is a must-have currently so I am approving.

File: distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli (hunk 5 above)
Comment (hmlnarik) [reply distribution/server-dist/src/main/docs/examples/map-storage-concurrenthashmap.cli:16]: Given the file name (`map` storage), I would prefer keeping it as is.

**Comment by hmlnarik** [comment]
> Should we add something also to the documentation/release notes?

Thank you, this is a good point. The documentation needs to be there though once we add at least one implementation of map storage capable of clustering, and thus open it for user preview. As for now, when it is only for experts, would it make sense to keep it rather visible only to experts, and thus undocumented?

**Review by vramik — APPROVED** [review]
Thanks!

**Pull request merged**